### PR TITLE
Ajout d'une vue pour calculer la progression des utilisateurs sur les…

### PR DESCRIPTION
**Pull request by jdauphant**
… 30 derniers jours

## Etape
- [x] Test sur la base de prod

## Log
[log ChatGPT.log](https://github.com/tchapgouv/tchap-stats/files/11311600/log.ChatGPT.log)

### CODE REVIEW

File: scripts/tables.sql [hunk 1]
```diff
@@ -84,3 +84,29 @@ GROUP BY device_id, month, user_id, instance, domain, device_type, platform;
...
+    LEFT JOIN user_daily_visits ON user_daily_visits.visit_ts >= date_range.day - INTERVAL '30 days'
+    AND user_daily_visits.visit_ts < date_range.day + INTERVAL '1 day'
+  GROUP BY
+    date_range.day
+)
+SELECT
+  day,
+  unique_user_count
+FROM
+  unique_users_per_day;
```
Comment (jdauphant) [inline scripts/tables.sql:112]:
```suggestion
CREATE MATERIALIZED VIEW daily_unique_user_count AS
SELECT
  date_trunc('day', visit_ts) AS day,
  user_id,
  COUNT(*) AS visits_count,
  COUNT(CASE
    WHEN user_agent LIKE 'Mozilla%' THEN 1
    ELSE NULL
  END) AS web_visits_count,
  COUNT(CASE
    WHEN user_agent LIKE 'Tchap%Android%' OR
         user_agent LIKE 'RiotNSE/2%iOS%' OR
         user_agent LIKE 'RiotSharedExtension/2%iOS%' OR
         user_agent LIKE 'Tchap%iOS%' OR
         user_agent LIKE 'Riot%iOS' OR
         user_agent LIKE 'Riot%Android' OR
         user_agent LIKE 'Element%Android' OR
         user_agent LIKE 'Element%iOS'
    THEN 1
    ELSE NULL
  END) AS mobile_visits_count,
  COUNT(CASE
    WHEN user_agent NOT LIKE 'Mozilla%' AND
         user_agent NOT LIKE 'Tchap%Android%' AND
         user_agent NOT LIKE 'RiotNSE/2%iOS%' AND
         user_agent NOT LIKE 'RiotSharedExtension/2%iOS%' AND
         user_agent NOT LIKE 'Tchap%iOS%' AND
         user_agent NOT LIKE 'Riot%iOS' AND
         user_agent NOT LIKE 'Riot%Android' AND
         user_agent NOT LIKE 'Element%Android' AND
         user_agent NOT LIKE 'Element%iOS'
    THEN 1
    ELSE NULL
  END) AS other_visits_count
FROM
  user_daily_visits
GROUP BY
  date_trunc('day', visit_ts),
  user_id;

CREATE MATERIALIZED VIEW unique_user_daily_count_30d AS
WITH date_range AS (
  SELECT generate_series(
    CURRENT_DATE - INTERVAL '3 months',
    CURRENT_DATE,
    INTERVAL '1 day'
  )::date AS day
),
unique_users_per_day AS (
  SELECT
    date_range.day,
    COUNT(DISTINCT daily_unique_user_count.user_id) AS unique_user_count
  FROM
    date_range
    LEFT JOIN daily_unique_user_count ON daily_unique_user_count.day BETWEEN date_range.day - INTERVAL '30 days' AND date_range.day - INTERVAL '1 day'
  WHERE
    NOT (mobile_visits_count = 0 AND web_visits_count = 0 AND other_visits_count > 0)
  GROUP BY
    date_range.day
)
SELECT
  day,
  unique_user_count
FROM
  unique_users_per_day;
```

**Comment by jdauphant** [comment]
Model: GPT-4

### Prompt : 
CREATE TABLE IF NOT EXISTS user_daily_visits (
  user_id VARCHAR NOT NULL,
  device_id VARCHAR NOT NULL,
  visit_ts timestamp with time zone NOT NULL,
  user_agent VARCHAR NOT NULL,
  instance VARCHAR NOT NULL,
  domain VARCHAR NOT NULL
);

For each day for the last 3 months, I want to calculate on user_daily_visits the number of unique user_id of the last 30 days.
it should be a materialized view named unique_user_daily_count_30d

### Answer
To calculate the number of unique users for each day for the last 3 months and store it in a materialized view, you can use the following SQL:

sql
Copy code
CREATE MATERIALIZED VIEW unique_user_daily_count_30d AS
WITH date_range AS (
  SELECT generate_series(
    CURRENT_DATE - INTERVAL '3 months',
    CURRENT_DATE,
    INTERVAL '1 day'
  )::date AS day
),
unique_users_per_day AS (
  SELECT
    date_range.day,
    COUNT(DISTINCT user_daily_visits.user_id) AS unique_user_count
  FROM
    date_range
    LEFT JOIN user_daily_visits ON user_daily_visits.visit_ts >= date_range.day - INTERVAL '30 days'
    AND user_daily_visits.visit_ts < date_range.day + INTERVAL '1 day'
  GROUP BY
    date_range.day
)
SELECT
  day,
  unique_user_count
FROM
  unique_users_per_day;
This query first generates a list of dates for the last 3 months (date_range), then counts the number of unique user_ids for each date in the range by joining the user_daily_visits table. Finally, it selects the day and unique_user_count columns from the unique_users_per_day CTE and stores the result in a materialized view called unique_user_daily_count_30d.


### Reponse prompt

Can you generate a materilized view that use less memory (user_daily_visits have millions of row) ?

### Response GPT-4

To create a more memory-efficient materialized view, you can use incremental aggregation. In this approach, you'll store the daily unique users count, and then use this data to calculate the rolling 30-day unique users count. This will help reduce the amount of memory used during the calculation.

First, create a materialized view to store the daily unique user count:

sql
Copy code
CREATE MATERIALIZED VIEW daily_unique_user_count AS
SELECT
  date_trunc('day', visit_ts) AS day,
  user_id,
  COUNT(*) AS visits_count
FROM
  user_daily_visits
GROUP BY
  date_trunc('day', visit_ts),
  user_id;
Now, create the materialized view for the rolling 30-day unique user count:

sql
Copy code
CREATE MATERIALIZED VIEW unique_user_daily_count_30d AS
WITH date_range AS (
  SELECT generate_series(
    CURRENT_DATE - INTERVAL '3 months',
    CURRENT_DATE,
    INTERVAL '1 day'
  )::date AS day
),
unique_users_per_day AS (
  SELECT
    date_range.day,
    COUNT(DISTINCT daily_unique_user_count.user_id) AS unique_user_count
  FROM
    date_range
    LEFT JOIN daily_unique_user_count ON daily_unique_user_count.day BETWEEN date_range.day - INTERVAL '30 days' AND date_range.day - INTERVAL '1 day'
  GROUP BY
    date_range.day
)
SELECT
  day,
  unique_user_count
FROM
  unique_users_per_day;
This approach stores the daily unique user count separately, reducing the amount of memory required for calculating the rolling 30-day unique user count. The result is stored in a materialized view named unique_user_daily_count_30d.

Keep in mind that you'll need to refresh both materialized views periodically to ensure they're up-to-date. You can do this with the REFRESH MATERIALIZED VIEW command:

sql
Copy code
REFRESH MATERIALIZED VIEW daily_unique_user_count;
REFRESH MATERIALIZED VIEW unique_user_daily_count_30d;

### Response prompt

in daily_unique_user_count can you had count of web, mobile, and other :
CASE
WHEN user_agent LIKE 'Mozilla%' THEN 'Web'
WHEN user_agent LIKE 'Tchap%Android%' THEN 'Mobile'
WHEN user_agent LIKE 'RiotNSE/2%iOS%' THEN 'Mobile'
WHEN user_agent LIKE 'RiotSharedExtension/2%iOS%' THEN 'Mobile'
WHEN user_agent LIKE 'Tchap%iOS%' THEN 'Mobile'
WHEN user_agent LIKE 'Riot%iOS' THEN 'Mobile'
WHEN user_agent LIKE 'Riot%Android' THEN 'Mobile'
WHEN user_agent LIKE 'Element%Android' THEN 'Mobile'
WHEN user_agent LIKE 'Element%iOS' THEN 'Mobile'
ELSE 'Other'
the corresponding column to had are mobile_visits_count, web_visits_count, other_visits_count

Also, unique_user_daily_count_30d should not include visits where mobile_visits_count =0 , web_visits_count =0 and other_visits_count > 0

### Response GPT-4
You can modify the daily_unique_user_count materialized view to include the count of web, mobile, and other visits by using the CASE statement and conditional aggregation. Additionally, you can update the unique_user_daily_count_30d view to exclude visits where mobile_visits_count = 0, web_visits_count = 0, and other_visits_count > 0.

First, create the updated daily_unique_user_count materialized view:

sql
Copy code
CREATE MATERIALIZED VIEW daily_unique_user_count AS
SELECT
  date_trunc('day', visit_ts) AS day,
  user_id,
  COUNT(*) AS visits_count,
  COUNT(CASE
    WHEN user_agent LIKE 'Mozilla%' THEN 1
    ELSE NULL
  END) AS web_visits_count,
  COUNT(CASE
    WHEN user_agent LIKE 'Tchap%Android%' OR
         user_agent LIKE 'RiotNSE/2%iOS%' OR
         user_agent LIKE 'RiotSharedExtension/2%iOS%' OR
         user_agent LIKE 'Tchap%iOS%' OR
         user_agent LIKE 'Riot%iOS' OR
         user_agent LIKE 'Riot%Android' OR
         user_agent LIKE 'Element%Android' OR
         user_agent LIKE 'Element%iOS'
    THEN 1
    ELSE NULL
  END) AS mobile_visits_count,
  COUNT(CASE
    WHEN user_agent NOT LIKE 'Mozilla%' AND
         user_agent NOT LIKE 'Tchap%Android%' AND
         user_agent NOT LIKE 'RiotNSE/2%iOS%' AND
         user_agent NOT LIKE 'RiotSharedExtension/2%iOS%' AND
         user_agent NOT LIKE 'Tchap%iOS%' AND
         user_agent NOT LIKE 'Riot%iOS' AND
         user_agent NOT LIKE 'Riot%Android' AND
         user_agent NOT LIKE 'Element%Android' AND
         user_agent NOT LIKE 'Element%iOS'
    THEN 1
    ELSE NULL
  END) AS other_visits_count
FROM
  user_daily_visits
GROUP BY
  date_trunc('day', visit_ts),
  user_id;
Next, create the updated unique_user_daily_count_30d materialized view:

sql
Copy code
CREATE MATERIALIZED VIEW unique_user_daily_count_30d AS
WITH date_range AS (
  SELECT generate_series(
    CURRENT_DATE - INTERVAL '3 months',
    CURRENT_DATE,
    INTERVAL '1 day'
  )::date AS day
),
unique_users_per_day AS (
  SELECT
    date_range.day,
    COUNT(DISTINCT daily_unique_user_count.user_id) AS unique_user_count
  FROM
    date_range
    LEFT JOIN daily_unique_user_count ON daily_unique_user_count.day BETWEEN date_range.day - INTERVAL '30 days' AND date_range.day - INTERVAL '1 day'
  WHERE
    NOT (mobile_visits_count = 0 AND web_visits_count = 0 AND other_visits_count > 0)
  GROUP BY
    date_range.day
)
SELECT
  day,
  unique_user_count
FROM
  unique_users_per_day;
The updated daily_unique_user_count materialized view includes the web_visits_count, mobile_visits_count, and other_visits_count columns. The unique_user_daily_count_30d materialized view now excludes visits where mobile_visits_count = 0, web_visits_count = 0, and other_visits_count > 0.

Remember to refresh both materialized views periodically to keep them up-to-date:

sql
Copy code
REFRESH MATERIALIZED VIEW daily_unique_user_count;
REFRESH MATERIALIZED VIEW unique_user

**Comment by jdauphant** [comment]
Les prompts sont donnée sur une version précédente

File: scripts/tables.sql [hunk 2]
```diff
@@ -84,3 +84,82 @@ GROUP BY device_id, month, user_id, instance, domain, device_type, platform;
...
+    THEN 1
+    ELSE NULL
+  END) AS other_visits_count
+FROM
+  user_daily_visits
+GROUP BY
+  date_trunc('day', visit_ts),
+  user_id,
+  instance,
+  domain;
```
Comment (jdauphant) [inline scripts/tables.sql:131]:
```suggestion
  domain;
  
CREATE INDEX idx_daily_unique_user_count_day ON daily_unique_user_count(day);
```

**Review by mcalinghee — APPROVED** [review]
Ok si cela a ete teste

**Pull request merged**